src/pkcs11/pkcs11-object.c

```diff
@@ -375,7 +375,7 @@ C_FindObjectsInit(CK_SESSION_HANDLE hSession,	/* the session's handle */
 
 	/* Check whether we should hide private objects */
 	hide_private = 0;
-	if (slot->login_user != CKU_USER && (slot->token_info.flags & CKF_LOGIN_REQUIRED))
+	if ((slot->login_user != CKU_SO) && (slot->login_user != CKU_USER) && (slot->token_info.flags & CKF_LOGIN_REQUIRED))
```

Comment:
There are 4 values the `slot->login_user` could have:
 * `-1` for not logged in user
 * `0` for `CKU_SO` login
 * `1` for `CKU_USER` login
 * `2` for `CKU_CONTEXT_SPECIFIC` (AFAIK this does not get propagated to the above variable, but is used for logins).

Your suggested change makes it very hard to understand what is going on there. Would it make sense to replace it with:

    if ((slot->login_user == -1) && (slot->token_info.flags & CKF_LOGIN_REQUIRED)

Reply:
Yes, you're probably right. It would be better to check against not logged in at all.